includes/enrolment/class-sensei-course-enrolment.php

```diff
@@ -444,4 +445,87 @@ public function reset_course_enrolment_salt() {
 
 		return $new_salt;
 	}
+
+	/**
+	 * Remove learner from the course, overriding the providers rule.
+	 *
+	 * @param int $user_id User ID.
+	 *
+	 * @return boolean Success flag.
+	 */
+	public function remove_learner( $user_id ) {
+		$removed_learners = $this->get_removed_learners();
+
+		if ( isset( $removed_learners[ $user_id ] ) ) {
+			return false;
+		}
+
+		$removed_learners[ $user_id ] = [ 'date' => time() ];
+
+		$this->save_enrolment( $user_id, false );
+
+		return $this->update_removed_learners( $removed_learners );
+	}
+
+	/**
+	 * Restore removed learner enrolment, giving the control back to the providers.
+	 *
+	 * @param int $user_id User ID.
+	 *
+	 * @return boolean Success flag.
+	 */
+	public function restore_removed_learner( $user_id ) {
+		$removed_learners = $this->get_removed_learners();
+
+		unset( $removed_learners[ $user_id ] );
+
+		return $this->update_removed_learners( $removed_learners );
+	}
+
+	/**
+	 * Check if the user is removed.
+	 *
+	 * @param int $user_id User ID.
+	 *
+	 * @return boolean Whether the learner is removed.
+	 */
+	public function check_removed_learner( $user_id ) {
+		$removed_learners = $this->get_removed_learners();
+
+		return in_array( $user_id, array_keys( $removed_learners ), true );
+	}
+
+	/**
+	 * Get removed learners meta.
+	 *
+	 * @return array Removed learners array.
+	 */
+	private function get_removed_learners() {
+		if ( isset( $this->removed_learners ) ) {
+			return $this->removed_learners;
+		}
+
+		$removed_learners_meta = get_post_meta( $this->course_id, self::META_REMOVED_LEARNERS, true );
+
+		if ( empty( $removed_learners_meta ) ) {
+			$this->removed_learners = [];
+		} else {
+			$this->removed_learners = json_decode( $removed_learners_meta, true );
+		}
+
+		return $this->removed_learners;
+	}
+
+	/**
+	 * Update removed learners meta.
+	 *
+	 * @param array $removed_learners Removed learners array.
+	 *
+	 * @return bool Whether it was updated.
+	 */
+	private function update_removed_learners( $removed_learners ) {
+		$this->removed_learners = $removed_learners;
+
+		return (bool) update_post_meta( $this->course_id, self::META_REMOVED_LEARNERS, wp_json_encode( $removed_learners ) );
```

Comment:
We should probably first update the meta and then set the value to the local variable. The way it is now if `update_post_meta` fails the two variables will be out of sync.

Reply:
Fixed in 582293f8e5e58b31dd0b46f041b8f85f8d703a14

Reply:
Good catch!